phoenix-core/src/main/java/org/apache/phoenix/jdbc/PhoenixStatement.java

```diff
@@ -371,6 +482,23 @@ private PhoenixResultSet executeQuery(final CompilableStatement stmt,
                                 plan =
                                         connection.getQueryServices().getOptimizer()
                                                 .optimize(PhoenixStatement.this, plan);
+                                setLastQueryPlan(plan);
+
+                                //verify metadata for the table/view/index in the query plan
+                                if (plan.getTableRef() != null && validateLastDdlTimestamp) {
+                                    try {
+                                        validateLastDDLTimestamp(plan.getTableRef(), true);
+                                    } catch (StaleMetadataCacheException e) {
+                                        String planSchemaName = plan.getTableRef().getTable().getSchemaName().toString();
+                                        String planTableName = plan.getTableRef().getTable().getTableName().toString();
+                                        // force update client metadata cache
+                                        LOGGER.debug("Force updating client metadata cache for {}", getInfoString(getLastQueryPlan().getTableRef()));
+                                        new MetaDataClient(connection).updateCache(connection.getTenantId(), planSchemaName, planTableName, true);
+                                        // skip last ddl timestamp validation in the retry
+                                        return executeQuery(stmt, doRetryOnMetaNotFoundError, queryLogger, noCommit, false);
```

Comment:
Isn't this still part of the outer try block and hence we will update the metrics twice? Cc @tkhurana

Reply:
@shahrs87 added a boolean flag for updating metrics in the finally block. Setting it to false in the catch blocks.